File: reddit_api/src/model/subreddit.rs
use std::borrow::Cow;
use std::collections::HashMap;

use crate::client::Client;
use crate::result::Result;
use crate::utils;
use crate::utils::{response_or_default, response_or_none_string};
use crate::{error::Error, model::flair::Flair};
use flutter_getter::FlutterGetters;
use flutter_rust_bridge::frb;
use getset::Getters;
use reqwest::Url;
use serde::{Deserialize, Serialize};
use serde_with::serde_as;
use serde_with::with_prefix;

use super::{Fullname, Thing};

with_prefix!(prefix_user_flair "user_flair_");

/// Subreddit's ID, is of the form `t5_xxx`
#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SubredditID(String);

#[derive(Default, Debug, Clone, PartialEq, Getters, FlutterGetters, Serialize, Deserialize)]
#[getset(get = "pub(crate)")]
pub struct SubredditInfo {
    /// The subreddit's name (e.g. "awww")
    #[serde(rename = "subreddit")]
    subreddit: String,
    #[serde(rename = "subreddit_id")]
    subreddit_id: SubredditID,
    /// The subreddit's name prefixed with "r/"
    #[serde(rename = "subreddit_name_prefixed")]
    subreddit_name_prefixed: String,
    /// The number of subscribers of the subreddit
    #[serde(rename = "subreddit_subscribers")]
    subscribers: u32,

    #[serde(rename = "sr_detail")]
    details: Option<Details>,
}

#[derive(Default, Debug, Clone, PartialEq, Getters, FlutterGetters, Serialize, Deserialize)]
#[getset(get = "pub(crate)")]
pub struct Common {
    #[serde(default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    user_is_contributor: bool,
    #[serde(rename = "banner_img", deserialize_with = "response_or_none_string")]
    banner_img: Option<String>,
    #[serde(rename = "allowed_media_in_comments", default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    allowed_media_in_comments: Vec<String>,
    #[serde(
        rename = "community_icon",
        deserialize_with = "response_or_none_string"
    )]
    #[getset(skip)]
    #[flutter_getter(skip)]
    community_icon: Option<String>,
    #[serde(rename = "user_is_banned", default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    user_is_banned: bool,
    #[serde(rename = "user_is_subscriber", default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    user_is_subscriber: bool,
    #[serde(rename = "free_form_reports", default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    free_form_reports: bool,
    #[serde(rename = "display_name")]
    display_name: String,
    #[serde(rename = "header_img")]
    header_img: Option<String>,
    #[serde(rename = "title", deserialize_with = "response_or_none_string")]
    title: Option<String>,
    #[serde(rename = "icon_size", deserialize_with = "response_or_default")]
    #[getset(skip)]
    #[flutter_getter(skip)]
    icon_size: [usize; 2],
    #[serde(rename = "icon_img", deserialize_with = "response_or_none_string")]
    #[getset(skip)]
    #[flutter_getter(skip)]
    icon_img: Option<String>,
    #[serde(rename = "primary_color", deserialize_with = "response_or_none_string")]
    primary_color: Option<String>,

    #[serde(rename = "subscribers", deserialize_with = "response_or_default")]
    subscribers: i64,
    name: Fullname,
    #[serde(default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    quarantine: bool,

    #[serde(rename = "display_name_prefixed")]
    display_name_prefixed: String,

    #[serde(rename = "user_is_muted", default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    user_is_muted: bool,

    #[serde(rename = "restrict_posting", default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    restrict_posting: bool,

    #[serde(rename = "key_color", deserialize_with = "response_or_none_string")]
    key_color: Option<String>,

    #[serde(default)]
    // missing on user subreddit
    created: f64,

    #[serde(
        rename = "submit_text_label",
        deserialize_with = "response_or_none_string"
    )]
    submit_text_label: Option<String>,

    #[serde(rename = "show_media", default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    show_media: bool,

    #[serde(rename = "public_description")]
    public_description: Option<String>,

    #[serde(rename = "banner_size")]
    banner_size: Option<Vec<i64>>,

    #[serde(rename = "created_utc", default)]
    // missing on user subreddit
    created_utc: f64,

    url: String,

    #[serde(rename = "restrict_commenting", default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    restrict_commenting: bool,

    #[serde(rename = "submit_link_label")]
    submit_link_label: Option<String>,

    #[serde(rename = "link_flair_enabled", default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    link_flair_enabled: bool,
    #[serde(rename = "link_flair_position")]
    link_flair_position: Option<String>,

    #[serde(rename = "header_size")]
    header_size: Option<Vec<i64>>,

    #[serde(rename = "accept_followers", default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    accept_followers: bool,

    #[serde(rename = "disable_contributor_requests", default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    disable_contributor_requests: bool,

    #[serde(rename = "subreddit_type")]
    subreddit_type: Option<String>,

    #[serde(rename = "user_is_moderator", default)]
    #[serde(deserialize_with = "utils::response_or_default")]
    user_is_moderator: bool,
}

#[derive(Default, Debug, Clone, PartialEq, Getters, FlutterGetters, Serialize, Deserialize)]
#[getset(get = "pub(crate)")]
pub struct Details {
    #[serde(default, deserialize_with = "utils::response_or_default")]
    default_set: bool,
    #[serde(default, deserialize_with = "utils::response_or_none_string")]
    icon_color: Option<String>,
    previous_names: Vec<String>,
    #[serde(default, deserialize_with = "utils::response_or_default")]
    over_18: bool,
    description: Option<String>,
    #[serde(flatten)]
    other: Common,
}

#[derive(Default, Debug, Clone, Getters, FlutterGetters, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[serde_as]
#[getset(get = "pub(crate)")]
pub struct Subreddit {
    #[serde(flatten)]
    other: Common,

    #[serde(flatten, with = "prefix_user_flair")]
    user_flair: Flair,
    #[serde(rename = "submit_text_html")]
    submit_text_html: Option<String>,
    #[serde(rename = "wiki_enabled")]
    wiki_enabled: Option<bool>,
    #[serde(rename = "user_can_flair_in_sr")]
    user_can_flair_in_sr: Option<bool>,
    #[serde(rename = "allow_galleries", deserialize_with = "response_or_default")]
    allow_galleries: bool,

    #[serde(
        rename = "active_user_count",
        deserialize_with = "response_or_default",
        default
    )]
    #[deprecated(note = "Reddit seems to have remove it from it responses")]
    active_user_count: u32,
    #[serde(
        rename = "accounts_active",
        deserialize_with = "response_or_default",
        default
    )]
    #[deprecated(note = "Reddit seems to have remove it from it responses")]
    accounts_active: u32,
    #[serde(rename = "public_traffic", deserialize_with = "response_or_default")]
    public_traffic: bool,
    #[serde(rename = "hide_ads", deserialize_with = "response_or_default")]
    hide_ads: bool,
    #[serde(
        rename = "prediction_leaderboard_entry_type",
        deserialize_with = "response_or_default"
    )]
    prediction_leaderboard_entry_type: i64,
    #[serde(rename = "emojis_enabled")]
    emojis_enabled: bool,
    #[serde(rename = "advertiser_category")]
    advertiser_category: Option<String>,
    #[serde(rename = "comment_score_hide_mins")]
    comment_score_hide_mins: Option<i64>,
    #[serde(rename = "allow_predictions", deserialize_with = "response_or_default")]
    allow_predictions: bool,
    #[serde(
        rename = "user_has_favorited",
        deserialize_with = "response_or_default"
    )]
    user_has_favorited: bool,
    #[serde(rename = "banner_background_image")]
    banner_background_image: Option<String>,
    #[serde(
        rename = "original_content_tag_enabled",
        deserialize_with = "response_or_default"
    )]
    original_content_tag_enabled: bool,
    #[serde(
        rename = "community_reviewed",
        deserialize_with = "response_or_default"
    )]
    community_reviewed: bool,
    #[serde(rename = "submit_text")]
    submit_text: Option<String>,
    #[serde(rename = "description_html")]
    description_html: Option<String>,
    #[serde(rename = "spoilers_enabled", deserialize_with = "response_or_default")]
    spoilers_enabled: bool,
    #[serde(rename = "comment_contribution_settings")]
    comment_contribution_settings: CommentContributionSettings,
    #[serde(rename = "allow_talks", deserialize_with = "response_or_default")]
    allow_talks: bool,
    #[serde(
        rename = "all_original_content",
        deserialize_with = "response_or_default"
    )]
    all_original_content: bool,
    #[serde(rename = "has_menu_widget", deserialize_with = "response_or_default")]
    has_menu_widget: bool,
    #[serde(rename = "is_enrolled_in_new_modmail")]
    is_enrolled_in_new_modmail: Option<bool>,
    #[serde(
        rename = "can_assign_user_flair",
        deserialize_with = "response_or_default"
    )]
    can_assign_user_flair: bool,
    wls: Option<u32>,
    #[serde(
        rename = "show_media_preview",
        deserialize_with = "response_or_default"
    )]
    show_media_preview: bool,
    #[serde(
        rename = "submission_type",
        deserialize_with = "response_or_none_string"
    )]
    submission_type: Option<String>,
    #[serde(rename = "allow_videogifs", deserialize_with = "response_or_default")]
    allow_videogifs: bool,
    #[serde(
        rename = "should_archive_posts",
        deserialize_with = "response_or_default"
    )]
    should_archive_posts: bool,
    #[serde(rename = "allow_polls", deserialize_with = "response_or_default")]
    allow_polls: bool,
    #[serde(
        rename = "collapse_deleted_comments",
        deserialize_with = "response_or_default"
    )]
    collapse_deleted_comments: bool,
    #[serde(rename = "emojis_custom_size")]
    emojis_custom_size: Option<Vec<i64>>,
    #[serde(rename = "public_description_html")]
    public_description_html: Option<String>,
    #[serde(rename = "allow_videos", deserialize_with = "response_or_default")]
    allow_videos: bool,
    #[serde(rename = "is_crosspostable_subreddit")]
    is_crosspostable_subreddit: Option<bool>,
    #[serde(rename = "notification_level", default)]
    notification_level: Option<String>,
    #[serde(
        rename = "should_show_media_in_comments_setting",
        deserialize_with = "response_or_default"
    )]
    should_show_media_in_comments_setting: bool,
    #[serde(
        rename = "can_assign_link_flair",
        deserialize_with = "response_or_default"
    )]
    can_assign_link_flair: bool,
    #[serde(
        rename = "allow_prediction_contributors",
        deserialize_with = "response_or_default"
    )]
    allow_prediction_contributors: bool,
    #[serde(rename = "user_sr_flair_enabled")]
    user_sr_flair_enabled: Option<bool>,
    #[serde(
        rename = "user_flair_enabled_in_sr",
        deserialize_with = "response_or_default"
    )]
    user_flair_enabled_in_sr: bool,
    #[serde(rename = "allow_discovery", deserialize_with = "response_or_default")]
    allow_discovery: bool,
    #[serde(
        rename = "user_sr_theme_enabled",
        deserialize_with = "response_or_default"
    )]
    user_sr_theme_enabled: bool,
    #[serde(rename = "suggested_comment_sort")]
    suggested_comment_sort: Option<String>,
    #[serde(
        rename = "banner_background_color",
        deserialize_with = "response_or_none_string"
    )]
    banner_background_color: Option<String>,
    id: String,
    #[serde(default, deserialize_with = "utils::response_or_default")]
    over18: bool,
    #[serde(rename = "header_title", deserialize_with = "response_or_none_string")]
    header_title: Option<String>,
    #[serde(rename = "description", deserialize_with = "response_or_none_string")]
    description: Option<String>,
    #[serde(rename = "allow_images", deserialize_with = "response_or_default")]
    allow_images: bool,
    #[serde(rename = "lang", deserialize_with = "response_or_none_string")]
    lang: Option<String>,
    #[serde(
        rename = "mobile_banner_image",
        deserialize_with = "response_or_none_string"
    )]
    mobile_banner_image: Option<String>,
    // #[serde(rename = "user_is_contributor")]
    // pub user_is_contributor: bool,
    #[serde(
        rename = "allow_predictions_tournament",
        deserialize_with = "response_or_default"
    )]
    allow_predictions_tournament: bool,
}

impl Subreddit {
    #[must_use]
    #[frb(sync, getter)]
    pub fn icon(&self) -> SubredditIcon {
        self.other.icon()
    }
    pub async fn subscribe(&mut self, client: &Client) -> Result<()> {
        client.subscribe(self.other.name()).await?;
        self.other.user_is_subscriber = true;
        Ok(())
    }
    pub async fn unsubscribe(&mut self, client: &Client) -> Result<()> {
        client.unsubscribe(self.other.name()).await?;
        self.other.user_is_subscriber = false;
        Ok(())
    }
    pub async fn favorite(&mut self, favorite: bool, client: &Client) -> Result<()> {
        client.favorite(self.other.display_name(), favorite).await?;
        self.user_has_favorited = favorite;
        Ok(())
    }
}

#[frb(non_opaque)]
pub enum SubredditIcon {
    Image(Icon),
    Color(String),
}

pub struct Icon {
    pub url: String,
    pub width: usize,
    pub height: usize,
}

impl Common {
    #[must_use]
    #[frb(sync, getter)]
    pub fn icon(&self) -> SubredditIcon {
        if let Some(url) = &self.community_icon {
            let url = Url::parse(url).expect("Could not parse community icon url");
            let params: HashMap<Cow<str>, Cow<str>> = url.query_pairs().collect();
            let width: usize = params
                .get("width")
                .expect("Width not found in community icon url")
                .parse()
                .expect("Could not parse width in community icon url");
            let height: usize = width;

            SubredditIcon::Image(Icon {
                url: url.to_string(),
                width,
                height,
            })
        } else if let Some(url) = &self.icon_img {
            SubredditIcon::Image(Icon {
                url: url.to_owned(),
                width: self.icon_size[0],
                height: self.icon_size[1],
            })
        } else {
            let color = self
                .key_color
                .clone()
                .unwrap_or(self.primary_color.clone().unwrap_or("black".to_owned()));
            SubredditIcon::Color(color)
        }
    }
}

impl Details {
    #[must_use]
    #[frb(sync, getter)]
    pub fn icon(&self) -> SubredditIcon {
        self.other.icon()
    }
}

impl TryFrom<Thing> for Subreddit {
    type Error = Error;

    fn try_from(value: Thing) -> Result<Self> {
        match value {
            Thing::Subreddit(subreddit) => Ok(subreddit),
            _ => Err(Error::InvalidThing),
        }
    }
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CommentContributionSettings {
    #[serde(
        rename = "allowed_media_types",
        deserialize_with = "response_or_default",
        default
    )]
    pub allowed_media_types: Vec<String>,
}

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum NotificationLevel {
    #[default]
    Unknown,
    Low,
}

File: app/src/main/java/com/sofamaniac/crabir/data/remote/dto/subreddit/subreddit.kt
/*
 * *
 *  * Created by sofamaniac
 *  * Copyright (c) 2026 . All rights reserved.
 *  * Last modified 1/12/26, 10:44 PM
 *
 */

package com.sofamaniac.crabir.data.remote.dto.subreddit

import com.sofamaniac.crabir.domain.model.Fullname
import com.sofamaniac.crabir.domain.model.SubredditData
import com.sofamaniac.crabir.domain.repository.DataInterface
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import tech.mappie.api.ObjectMappie

@Serializable
@JvmInline
value class SubredditId(val id: String)

@Serializable
data class CommentContributionSettings(
    @SerialName("allowed_media_types") val allowedMediaTypes: List<String>? = null
)

fun dummySubredditData(): SubredditDTO {
    return SubredditDTO(
        displayName = "",
        id = "",
        iconImg = "",
        primaryColor = "black",
        keyColor = "black",
        name = Fullname(""),
    )
}
// TODO sometimes, id, display_name and icon_img are missing

@Serializable
data class SubredditDTO(
    @SerialName("user_flair_background_color")
    val userFlairBackgroundColor: String? = null,
    @SerialName("submit_text_html")
    val submitTextHtml: String? = null,
    @SerialName("user_is_banned")
    val userIsBanned: Boolean = false,
    @SerialName("free_form_reports")
    val freeFormReports: Boolean = false,
    @SerialName("wiki_enabled")
    val wikiEnabled: Boolean? = null,
    @SerialName("user_is_muted")
    val userIsMuted: Boolean = false,
    @SerialName("user_can_flair_in_sr")
    val userCanFlairInSr: Boolean? = null,
    /** The name of the subreddit (e.g. "unixporn") */
    @SerialName("display_name")
    val displayName: String = "",
    @SerialName("title")
    val title: String = "",
    @SerialName("allow_galleries")
    val allowGalleries: Boolean = false,
    @SerialName("icon_size")
    val iconSize: List<Int>? = null,
    /** Primary color in the form "#ffffff" */
    @SerialName("primary_color")
    val primaryColor: String? = null,
    @SerialName("active_user_count")
    val activeUserCount: Int? = null,
    @SerialName("icon_img")
    val iconImg: String? = null,
    /** The name of the subreddit with the 'r/' prefix (e.g. "r/unixporn") */
    @SerialName("display_name_prefixed")
    val displayNamePrefixed: String = "",
    @SerialName("accounts_active")
    val accountsActive: Int? = null,
    @SerialName("public_traffic")
    val publicTraffic: Boolean = false,
    @SerialName("subscribers")
    val subscribers: Int = 0,
    @SerialName("videostream_links_count")
    val videostreamLinksCount: Int = 0,
    /** The string "t2_[id]" */
    @SerialName("name")
    override val name: Fullname,
    @SerialName("quarantine")
    val quarantine: Boolean = false,
    @SerialName("hide_ads")
    val hideAds: Boolean = false,
    @SerialName("prediction_leaderboard_entry_type")
    val predictionLeaderboardEntryType: Int = 0,
    @SerialName("emojis_enabled")
    val emojisEnabled: Boolean = false,
    @SerialName("advertiser_category")
    val advertiserCategory: String = "",
    @SerialName("public_description")
    val publicDescription: String = "",
    @SerialName("comment_score_hide_mins")
    val commentScoreHideMins: Int = 0,
    @SerialName("allow_predictions")
    val allowPredictions: Boolean = false,
    /** Escaped HTML url */
    @SerialName("community_icon")
    val communityIcon: String? = null,
    @SerialName("original_content_tag_enabled")
    val originalContentTagEnabled: Boolean = false,
    @SerialName("community_reviewed")
    val communityReviewed: Boolean = false,
    @SerialName("description_html")
    val descriptionHtml: String? = null,
    @SerialName("spoilers_enabled")
    val spoilersEnabled: Boolean = false,
    @SerialName("comment_contribution_settings")
    val commentContributionSettings: CommentContributionSettings? = null,
    @SerialName("allow_talks")
    val allowTalks: Boolean = false,
    @SerialName("user_flair_position")
    val userFlairPosition: String? = null,
    @SerialName("all_original_content")
    val allOriginalContent: Boolean = false,
    @SerialName("has_menu_widget")
    val hasMenuWidget: Boolean = false,
    @SerialName("key_color")
    val keyColor: String = "",
    @SerialName("can_assign_user_flair")
    val canAssignUserFlair: Boolean = false,
    @SerialName("created")
    val created: Double = 0.0,
    @SerialName("show_media_preview")
    val showMediaPreview: Boolean = false,
    @SerialName("submission_type")
    val submissionType: String = "",
    @SerialName("user_is_subscriber")
    val userIsSubscriber: Boolean = false,
    @SerialName("user_has_favorited")
    val userHasFavorited: Boolean = false,
    @SerialName("allowed_media_in_comments")
    val allowedMediaInComments: List<String> = emptyList(),
    @SerialName("allow_videogifs")
    val allowVideogifs: Boolean = false,
    @SerialName("should_archive_posts")
    val shouldArchivePosts: Boolean = false,
    @SerialName("user_flair_type")
    val userFlairType: String? = null,
    @SerialName("allow_polls")
    val allowPolls: Boolean = false,
    @SerialName("collapse_deleted_comments")
    val collapseDeletedComments: Boolean = false,
    @SerialName("emojis_custom_size")
    val emojisCustomSize: List<Int>? = null,
    @SerialName("public_description_html")
    val publicDescriptionHtml: String? = null,
    @SerialName("allow_videos")
    val allowVideos: Boolean = false,
    @SerialName("notification_level")
    val notificationLevel: String? = null,
    @SerialName("should_show_media_in_comments_setting")
    val shouldShowMediaInCommentsSetting: Boolean = false,
    @SerialName("can_assign_link_flair")
    val canAssignLinkFlair: Boolean = false,
    @SerialName("accounts_active_is_fuzzed")
    val accountsActiveIsFuzzed: Boolean = false,
    @SerialName("allow_prediction_contributors")
    val allowPredictionContributors: Boolean = false,
    @SerialName("link_flair_position")
    val linkFlairPosition: String = "",
    @SerialName("user_sr_flair_enabled")
    val userSrFlairEnabled: Boolean? = null,
    @SerialName("user_flair_enabled_in_sr")
    val userFlairEnabledInSr: Boolean = false,
    @SerialName("allow_discovery")
    val allowDiscovery: Boolean = false,
    @SerialName("accept_followers")
    val acceptFollowers: Boolean = false,
    @SerialName("user_sr_theme_enabled")
    val userSrThemeEnabled: Boolean = false,
    @SerialName("link_flair_enabled")
    val linkFlairEnabled: Boolean = false,
    @SerialName("disable_contributor_requests")
    val disableContributorRequests: Boolean = false,
    @SerialName("subreddit_type")
    val subredditType: String = "",

    // BANNER
    @SerialName("banner_background_image")
    val bannerBackgroundImage: String? = null,
    @SerialName("banner_img")
    val bannerImg: String = "",
    @SerialName("banner_background_color")
    val bannerBackgroundColor: String? = null,
    @SerialName("banner_size")
    val bannerSize: List<Int>? = null,
    @SerialName("mobile_banner_image")
    val mobileBannerImage: String? = null,


    @SerialName("show_media")
    val showMedia: Boolean = false,
    @SerialName("user_is_moderator")
    val userIsModerator: Boolean = false,
    @SerialName("over18")
    val over18: Boolean = false,

    // HEADER
    @SerialName("header_img")
    val headerImg: String? = null,
    @SerialName("header_title")
    val headerTitle: String? = null,
    @SerialName("header_size")
    val headerSize: List<Int>? = null,

    // SUBMIT TEXT
    @SerialName("submit_text_label")
    val submitTextLabel: String = "",
    @SerialName("submit_text")
    val submitText: String = "",
    @SerialName("submit_link_label")
    val submitLinkLabel: String = "",

    @SerialName("description")
    val description: String = "",
    @SerialName("allow_images")
    val allowImages: Boolean = false,
    @SerialName("lang")
    val lang: String = "",
    @SerialName("url")
    val url: String = "",
    @SerialName("created_utc")
    val createdUtc: Double = 0.0,
    @SerialName("user_is_contributor")
    val userIsContributor: Boolean = false,
    @SerialName("allow_predictions_tournament")
    val allowPredictionsTournament: Boolean = false,
    @SerialName("id")
    override val id: String,
    @SerialName("restrict_posting")
    val restrictPosting: Boolean = false,
    @SerialName("restrict_commenting")
    val restrictCommenting: Boolean = false,
) : DataInterface {
    val icon: SubredditIcon =
        when {
            !communityIcon.isNullOrBlank() -> SubredditIcon.Icon(communityIcon)
            !iconImg.isNullOrBlank() -> SubredditIcon.Icon(iconImg)
            keyColor.isNotBlank() -> SubredditIcon.Color(keyColor)
            !primaryColor.isNullOrBlank() -> SubredditIcon.Color(primaryColor)
            else -> SubredditIcon.Color("black")
        }
}


object SubredditDetailsMapper : ObjectMappie<SubredditDTO, SubredditData>() {
    override fun map(from: SubredditDTO): SubredditData = mapping {
        SubredditData::defaultSet fromValue false
        SubredditData::previousNames fromValue emptyList()
        SubredditData::iconColor fromProperty from::keyColor
    }
}

